### ISSUE TITLE
Automation for PR?

### ISSUE THREAD

**Issue by iranzo** [Pelican-Elegant/elegant#191]
Some services like https://mergify.io/ allow to automate PR workflow once they have LGTM, etc is this something we should consider?

**Comment by talha131** [comment]
Also if we can setup a service that formats the code and check coding conventions.

**Comment by iranzo** [comment]
If we have a tool to check the format we could.

For python I used travis with 'tox' which automates setup of virtual environment to install tools and run  things like:

- markdownlinter
- flake8, etc

If all of them finishes ok, it gives ok to that 'test'

As an Open Source project we do have access to travis 'free' service to do this if we identify the tools.

**Comment by talha131** [comment]
Here are the notes from my personal todo list. Just notes. I haven't used any of them yet.

# Visual Regression Testing

I wanted to use https://github.com/argos-ci/argos from the creators of material-ui. But it seems it is a WIP. Documentation is lacking.

# Code quality 

1. https://bettercodehub.com/
1. https://codeclimate.com/quality/
1. https://houndci.com/
1. https://stickler-ci.com/
1. https://lgtm.com/

We can also set up a travis task to run [prettier](https://github.com/prettier/prettier) on code. I am using it with my JS and TS project. But I am not sure how a travis task will update the code for us. What we want is to run the tests and format the code when a PR is merged or a change is pushed to the master.

**Comment by iranzo** [comment]
I do use codeclimate on some of them, and it's really easy to setup (similar to others) there are tools that can edit and submit code automatically, but I do usually prefer to have checks, report what is bad and tell PR author to fix them.

Similars have 'codacy',  (example of one project <https://app.codacy.com/project/citellus/citellus/dashboard> )

**Comment by iranzo** [comment]
![image](https://user-images.githubusercontent.com/312463/49565199-a6f5cc00-f926-11e8-961c-cfeed27c5697.png)

Kind of extra feedback we would get on new PR's

**Comment by iranzo** [comment]
I'm evaluating some of the urls you provided to see what could make sense

**Comment by iranzo** [comment]
From my test, Codacy, LGTM look good, from stickler-ci and houndci nothing yet

LGTM test over the repo (could add it externally for analysis as others require to install into the github organization for Pelican)

<https://lgtm.com/projects/g/Pelican-Elegant/elegant/alerts?mode=list>

**Comment by silverhook** [comment]
I think it’s a good idea to keep the community up to speed.

As Elegant is a theme (and not code that’s able to destroy original data), there is low risk in moving forward quickly and adopting the “release early, release often” mantra.

Regarding which of the solutions would fit us best, I have no experience with any of them.

@calfzhou, any concerns?

**Comment by iranzo** [comment]
I'm on the 'release early, release often'  approach :-) of course, we should determine what we do put in 'master' or in devel branches and how that is going to be approached. Probably doing master only + 'release tags' will be easier (not sure if that was already discussed).

But in this topic, whatever we put as PR automation can be tested and rolled back if not covering the needs

**Comment by calfzhou** [comment]
`release early, release often` is suitable for this project.

**Comment by iranzo** [comment]
About the 'bots' that were checked?, do you want me to setup any of those in the 'elegant' repo?
 (tested codacy and lgtm with better impression to me), also this was raised as per 'mergify' in terms of putting new commits on merge automatically once they have been reviewed.

What would be the project goal here?

**Comment by iranzo** [comment]
With validate commits, we do have now a simple test ran for each execution (next step will be to trigger documentation update when commit is merged).

In parallel I've started adding codacy ( <https://app.codacy.com/project/Elegant/elegant/dashboard> ) and once is reviewed, checked, we can start with LGTM

**Comment by iranzo** [comment]
BTW, with travis we now have some 'warning' while the build hasn't yet succeeded:


![image](https://user-images.githubusercontent.com/312463/49790084-acd81c80-fd2d-11e8-8082-900c9c4c27bd.png)


Regards,
Pablo

**Comment by talha131** [comment]
@iranzo can we (should we) have some bot that formats code whenever a patch is made?

1. Format html, CSS and JS
1. Format CSS and use our convention (https://github.com/Pelican-Elegant/elegant/blob/master/CONTRIBUTING.md#code-style)
1. Or at least report issues if it does not follow the conventions

Ideally it would be better if it can do formatting itself. Because if it will report issues but not fix them, then this will release in increasing friction to the new contributions. Potential contributors will have to install additional software to format the code and re-submit the patch.

On the other hand, auto format by a bot may create problems for us down the line.

What do you think? Is it possible and is it worth having?

**Comment by iranzo** [comment]
Can you check 'hound' was one of the suggested but seems not to pick elegant repo with my github account.

You can check at <https://houndci.com/> and enabling for Pelican-Elegant organization

**Comment by iranzo** [comment]
Now it shows, but we'll need to create a configuration file for each repo

**Comment by talha131** [comment]
@iranzo which issues houndci check for? I checked their site but they only thing I found was about trailing white spaces.

Can you please also look into https://bettercodehub.com/?

**Comment by iranzo** [comment]
> @iranzo which issues houndci check for? I checked their site but they only thing I found was about trailing white spaces.
> 
> Can you please also look into https://bettercodehub.com/?

Can you try adding it to organization (as I didn't enabled at first, it doesn't show elegant like hound did), once it's there, I can try it and then we decide and remove what we don't use

**Comment by talha131** [comment]
@iranzo please revoke your personal access to BCH in github settings. (Your personal, not the pelican-elegant org). Then try again. You should see elegant projects in BCH this time.

![](https://i.imgur.com/XEkguY8.png)

**Comment by iranzo** [comment]
I do see it now, thank!

**Comment by iranzo** [comment]
For hound: I've added a bunch of files to https://github.com/Pelican-Elegant/elegant/pull/246 but failed to get anything reported, so not sure if it's me, can someone else have another look on it?

**Comment by iranzo** [comment]
BTW in regards to hound, with update of Tipue to 7.0 it started to complain: https://github.com/Pelican-Elegant/elegant/pull/259 so apparently it works 'somehow'

**Comment by iranzo** [comment]
So, about current status, do we want PR's to report failed UT's for hound and w3c or should we just ignore them until we fix them for real?

(the one on hound and tipue is beyond our control as it's the file we ship from upstream, not our code)

**Comment by talha131** [comment]
> (the one on hound and tipue is beyond our control as it's the file we ship from upstream, not our code)

How about we make changes in tipue code to resolve hound issues? What do you think? Tipue is distributed with our theme, so it would make sense if we resolve issues in it. But on the other hand, tinkering with someone else code may require more time than we can give and it may not be the best use of our time.

**Comment by iranzo** [comment]
That's my concern, if we could submit patches to tipue it would be great so that next version they release, already contain them for others too, if we are going to work on 'forks' that will be extra load on us to mantain

So probably best way is to ask tipue.com to upload to repository and collaborate there or ask our CI's to ignore that subfolder...

**Comment by talha131** [comment]
> ask our CI's to ignore that subfolder...

Lets do it.

**Comment by iranzo** [comment]
Attempted in https://github.com/Pelican-Elegant/elegant/pull/264

**Comment by talha131** [comment]
@iranzo I am successfully using [pre-commit](https://pre-commit.com/) to auto-format the code with every commit in my other project. It also enables all the contributors of the said project to keep their configuration is sync using git.

I was wondering what if we add it Elegant. For example, we can use following [hooks](https://pre-commit.com/hooks.html)

1. jinja-lint
1. html-lint
1. prettier (to format our jinja code)

The advantage is this way all our code will always have the same format. (thanks to prettier).

Disadvantage is that this cannot be run online. If a user creates commit or PR using github online editor, then these hooks will not run. 

What are your thoughts on it?

**Comment by iranzo** [comment]
In worst case scenario, it can be done later and for users using git directly, it will be 'easier'.

+1

**Comment by talha131** [comment]
@iranzo What do you think of this tool
 
https://github.com/hiimbex/markdownify

**Comment by iranzo** [comment]
Looks good, however they won't render as Elegant articles, just easier
checking from or of new articles

Want me to activate ?

**Comment by talha131** [comment]
Yeah. It's better to activate. It would make reviewing PR easier.

**Comment by iranzo** [comment]
Enabled in documentation repo

**Comment by iranzo** [comment]
Changed to all repositories in the organization

**Comment by iranzo** [comment]
> @iranzo I am successfully using [pre-commit](https://pre-commit.com/) to auto-format the code with every commit in my other project. It also enables all the contributors of the said project to keep their configuration is sync using git.
> 
> I was wondering what if we add it Elegant. For example, we can use following [hooks](https://pre-commit.com/hooks.html)
> 
> 1. jinja-lint
> 2. html-lint
> 3. prettier (to format our jinja code)
> 
> The advantage is this way all our code will always have the same format. (thanks to prettier).
> 
> Disadvantage is that this cannot be run online. If a user creates commit or PR using github online editor, then these hooks will not run.
> 
> What are your thoughts on it?

Are you enabling it for the repo?

**Comment by talha131** [comment]
Yes. I have got in my local branch. Will push after releases.

**Comment by talha131** [comment]
@iranzo I tried pre-commit for Elegant. It has been a disappointing experience.

pre-commit hooks rely on external tools to lint or format code. If the external tool sucks or does not exist, pre-commit can't be of much help then.

In my day job, our code base is in C, C++, ObjC. We use clang-format hook from pre-commit. It has proved to be very useful and time saver.

For Elegant, our code base is in Jinja2. We need a Jinja2 linter and formatter. Unfortunately the one present, i.e., jinja-lint, is no where a finished product.

I tried using it. Some of its warnings are false positive. It gives incorrect warnings about `base.html`. A few of its warning does not making any sense like

> An inline parent element must only contain inline children

This warning is for `<head>`. What does it mean? Beats me.

When I tried to follow and fix its warning, it messed up html portion of the code.

Rest of the two that I mentioned, prettier and html-lint`, does not support Jinja2.

We direly need a formatter and linter for Jinaj2. jinja-lint falls behind by a large margin. Until it improves, there is no point in adding pre-commit.

**Comment by iranzo** [comment]
Ok :/ let's keep it without pre-commit for now then

**Comment by Seluj78** [comment]
Just for a heads up: https://github.com/motet-a/jinjalint/issues/16

**Comment by talha131** [comment]
We need spell check, especially for documentation site.

`Beginning` was misspelled as `beggining` since https://github.com/Pelican-Elegant/documentation/commit/84da2de

**Comment by talha131** [comment]
@iranzo Is it possible when a user opens a PR

1. We notify him to open it against next
1. Update the change log

Both of them are mandatory in most cases but not all. For example, it is possible to have a change that does not required updating change log. Or when we are doing a release, then PR has to be opened against master branch. 

What do you think?

If it is possible then should we do a simple reminder or block the PR until conditions are met?

**Comment by iranzo** [comment]
We can use that: 

https://help.github.com/en/articles/creating-a-pull-request-template-for-your-repository

**Comment by iranzo** [comment]
https://github.com/Pelican-Elegant/elegant/issues/353

**Comment by talha131** [comment]
@iranzo what are we using houndci for? So far it passes all the test. 

Let me know what it can do for us. If it serves no purpose then we should remove it.

I was going through their docs [here](http://help.houndci.com/collections/33942-configuration) and didn't see any linter that helps us.

I think our current setup is pretty good. Let me what you think.

**Comment by iranzo** [comment]
It only complained when the 'included' libraries were not excluded (tipue,
etc), so +1 to disabling it!
-- 

Pablo Iranzo Gómez

GnuPG: 0x5BD8E1E4
Web: https://iranzo.github.io  ||  https://citellus.org

**Comment by talha131** [comment]
Ok. Open the PR please.

I am closing this issue. It has served its purpose.

**Comment by talha131** [comment]
@iranzo we still have hound in our tests

![](https://i.imgur.com/LpszP21.png)

I think you need to remove this repository from your hound account.

**Comment by iranzo** [comment]
Let me do that too, thanks!

**Comment by iranzo** [comment]
I've tried to disable, let's see if it works now as it should!
